Project the root node of empty blocks

crisbeto opened the pull request:
Includes a couple of commits that allow for the root node of an `@empty` block to be content projected.

### fix(compiler): project empty block root node
Expands the workaround from https://github.com/angular/angular/pull/52414 to allow for the root nodes of `@empty` blocks to be content projected.

### fix(compiler): project empty block root node in template pipeline
Updates the template pipeline to allow for the root node of an `@empty` block to be content projected.

Fixes https://github.com/angular/angular/issues/53570.

crisbeto committed:
fix(compiler): project empty block root node

Expands the workaround from #52414 to allow for the root nodes of `@empty` blocks to be content projected.

Fixes #53570.

crisbeto commented on packages/compiler/src/template/pipeline/src/util/elements.ts:
```diff
@@ -20,6 +20,9 @@ export function createOpXrefMap(unit: CompilationUnit):
       continue;
     }
     map.set(op.xref, op);
+    if (op.kind === ir.OpKind.RepeaterCreate && op.emptyView !== null) {
```
It feels weird to be special-casing `RepeaterCreate` like this here and in `const_collection.ts`, but the problem is that the traits aren't really set up to allow for more than one reference per op. I'm open to suggestions.

dylhunn replied:
I have two refactors planned that will dramatically improve this situation:
1. We should automatically maintain an map of all ops with an `xref` field to the op instance, and update it "behind the scenes" when someone calls an OpList add/remove method. This will let us get rid of this element map helper function entirely.
2. We absolutely need to refactor the slot consumer trait to support arbitrarily many slots. (Same thing for the other traits.)

I assume you don't want to do either of these now, but please do put a comment here to remind me to hit this spot, especially for refactor #2.

crisbeto replied:
Done! Left a couple of `TODO(dylhunn)`.

pkozlowski-opensource reviewed (approved):
LGTM - this was fast!

dylhunn reviewed (approved):
You mentioned you were worried the template pipeline part was messy, but IMO it's not that messy! Just the one thing you already pointed out, which is a known refactoring task :)

crisbeto committed:
fix(compiler): project empty block root node in template pipeline

Updates the template pipeline to allow for the root node of an `@empty` block to be content projected.

atscott commented:
This PR was merged into the repository by commit 478d6222650884478314985e3d5132587c4f670c.
